sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala

```diff
@@ -74,10 +84,15 @@ case class Generate(
           } else {
             outputRows.map(or => joinProjection(joinedRow(row, or)))
           }
+        } ++ LazyIterator(() => boundGenerator.terminate()).map { row =>
+          joinedRow(nullOriginalRow, row)
         }
       }
     } else {
-      child.execute().mapPartitions(iter => iter.flatMap(row => boundGenerator.eval(row)))
+      child.execute().mapPartitions(iter =>
+        iter.flatMap(row => boundGenerator.eval(row)) ++
+          LazyIterator(() => boundGenerator.terminate())
```

Comment:
Looks like you are calling `terminate` on each partition. Is that not the same as how Hive does? In Hive it seems that `close` is called after all rows are processed.